### PR TITLE
treewide: migrate to the new rust fetcher

### PR DESCRIPTION
**Pull request by JohnRTitor**
Followup of https://github.com/NixOS/nixpkgs/pull/349360

CC: https://github.com/NixOS/nixpkgs/issues/327063

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by JohnRTitor** [comment]
I tried to migrate the cosmic packages like https://github.com/NixOS/nixpkgs/blob/e2b0cccf5508f35c234ab79dc29c471e60825d7c/pkgs/by-name/co/cosmic-applets/package.nix#L28 . But some of them failed, could you check @TomaSajt ?

<details>

```
       > Unpacking to /nix/store/2qzby0ikws1xcach4ckvlq4bi7268y4w-cosmic-applets-1.0.0-alpha.1-vendor/i18n-embed-fl-0.6.7
       > Unpacking to /nix/store/2qzby0ikws1xcach4ckvlq4bi7268y4w-cosmic-applets-1.0.0-alpha.1-vendor/i18n-embed-fl-0.8.0
       > Unpacking to /nix/store/2qzby0ikws1xcach4ckvlq4bi7268y4w-cosmic-applets-1.0.0-alpha.1-vendor/i18n-embed-impl-0.8.3
       > Unpacking to /nix/store/2qzby0ikws1xcach4ckvlq4bi7268y4w-cosmic-applets-1.0.0-alpha.1-vendor/iana-time-zone-0.1.60
       > Unpacking to /nix/store/2qzby0ikws1xcach4ckvlq4bi7268y4w-cosmic-applets-1.0.0-alpha.1-vendor/iana-time-zone-haiku-0.1.2
       > Traceback (most recent call last):
       >   File "/nix/store/d9ry6ifpm48hlsgs5xa6j1k1n11akbwf-fetch-cargo-vendor-util/bin/fetch-cargo-vendor-util", line 285, in <module>
       >     main()
       >   File "/nix/store/d9ry6ifpm48hlsgs5xa6j1k1n11akbwf-fetch-cargo-vendor-util/bin/fetch-cargo-vendor-util", line 281, in main
       >     subcommand_func()
       >   File "/nix/store/d9ry6ifpm48hlsgs5xa6j1k1n11akbwf-fetch-cargo-vendor-util/bin/fetch-cargo-vendor-util", line 273, in <lambda>
       >     "create-vendor": lambda: create_vendor(vendor_staging_dir=Path(sys.argv[2]), out_dir=Path(sys.argv[3]))
       >                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
       >   File "/nix/store/d9ry6ifpm48hlsgs5xa6j1k1n11akbwf-fetch-cargo-vendor-util/bin/fetch-cargo-vendor-util", line 230, in create_vendor
       >     copy_and_patch_git_crate_subtree(git_tree, pkg["name"], crate_out_dir)
       >   File "/nix/store/d9ry6ifpm48hlsgs5xa6j1k1n11akbwf-fetch-cargo-vendor-util/bin/fetch-cargo-vendor-util", line 167, in copy_and_patch_git_crate_subtree
       >     crate_manifest_path = find_crate_manifest_in_tree(git_tree, crate_name)
       >                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
       >   File "/nix/store/d9ry6ifpm48hlsgs5xa6j1k1n11akbwf-fetch-cargo-vendor-util/bin/fetch-cargo-vendor-util", line 163, in find_crate_manifest_in_tree
       >     raise Exception(f"Couldn't find manifest for crate {crate_name} inside {tree}.")
       > Exception: Couldn't find manifest for crate iced inside /nix/store/vqnd8w9d64flfvmmpi0l4z0p33098hpy-cosmic-applets-1.0.0-alpha.1-vendor-staging/git/4f77edd249e1c9cd525232050cec00c752ce7860.
       For full logs, run 'nix log /nix/store/ffcs6ds7s69rnq823r044h6ssxpd8fwb-cosmic-applets-1.0.0-alpha.1-vendor.drv'.
```

<details>

**Comment by JohnRTitor** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 356385`

---
### `x86_64-linux`
<details>
  <summary>:x: 2 packages failed to build:</summary>
  <ul>
    <li>chiptrack</li>
    <li>conduwuit</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 6 packages built:</summary>
  <ul>
    <li>alfis</li>
    <li>alfis-nogui</li>
    <li>cloud-hypervisor</li>
    <li>cloud-hypervisor.debug</li>
    <li>mullvad</li>
    <li>workstyle</li>
  </ul>
</details>

**Comment by TomaSajt** [comment]
> I tried to [migrate](https://github.com/JohnRTitor/nixpkgs/commit/0b2aaecd825984f50ab1ef9d57584786b948d5fa) the cosmic packages like
> 
> https://github.com/NixOS/nixpkgs/blob/e2b0cccf5508f35c234ab79dc29c471e60825d7c/pkgs/by-name/co/cosmic-applets/package.nix#L28
> . But some of them failed, could you check @TomaSajt ?

Seems like importCargoLock also fetches git submodules, but my fetchCargoVendor does not. Will look into it.

**Comment by TomaSajt** [comment]
Looks like an false assumption on my part: I assumed `fetchgit` has `fetchSubmodules` set to `false` as a default, just like `fetchFromGitHub`.
It looks like it is not the case: `fetchgit` has `fetchSubmodules = true` by default.

Thanks for reporting the issue, will make a PR.
I checked, there shouldn't be more differences between the defaults of `fetchgit` and how I'm calling `nix-prefetch-git`

**Comment by JohnRTitor** [comment]
@ofborg build alfis workstyle mullvad chiptrack cloud-hypervisor conduwuit

**Comment by TomaSajt** [comment]
@ofborg build cloud-hypervisor

**Comment by JohnRTitor** [comment]
> @ofborg build cloud-hypervisor

Looks like a hash mismatch for cargoHash, could you take care of it for me and push? Not near a computer right now.

**Comment by TomaSajt** [comment]
Consistently gets `sha256-7wIt4Ul65i0rqwGdsCJNicaU8Vkv/3KbLohc/ybXiEs=` locally and `sha256-xqMUB9aqkUIpnX0U30CfiWmjDI7IS5SuJIKF5byXIxk=` on ofborg...
very weird

**Comment by JohnRTitor** [comment]
Let's hold any mass migrations until this is resolved.

Very weird, only some packages are affected, in others the hash is rock solid.

**Comment by TomaSajt** [comment]
@ofborg build cloud-hypervisor

File: pkgs/build-support/rust/fetch-cargo-vendor-util.py [hunk 1]
```diff
@@ -119,7 +120,7 @@ def create_vendor_staging(lockfile_path: Path, out_dir: Path) -> None:
     shutil.copy(lockfile_path, out_dir / "Cargo.lock")
 
     # create a pool with at most 10 concurrent jobs
-    with mp.Pool(min(10, mp.cpu_count())) as pool:
+    with mp.Pool(min(1, mp.cpu_count())) as pool:
```
Comment (JohnRTitor) [inline pkgs/build-support/rust/fetch-cargo-vendor-util.py:123]: By the way I think you should tone this down. I got an error a while ago saying connection limits have exceeded. Happened only once though.

Comment (TomaSajt) [reply pkgs/build-support/rust/fetch-cargo-vendor-util.py:123]: Currently changed it to have the logs in order. Will consider

**Comment by TomaSajt** [comment]
I somehow reproduced the other hash locally, will do some testing to see what triggered it.

And yes, it seems like it is because of submodules.
The earlier hash did not fetch them, while the later one did.

**Comment by TomaSajt** [comment]
By the way, did some searching and it looks like cargo respects a setting that would make certain git deps not fetch their submodules:
https://github.com/rust-lang/cargo/pull/10717

I don't think we should bother to respect that. The more files we fetch, the less likely it is that the current implementation can't handle some edge case.

**Comment by niklaskorz** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 356385`

---
### `aarch64-darwin`
<details>
  <summary>:fast_forward: 3 packages marked as broken and skipped:</summary>
  <ul>
    <li>alfis</li>
    <li>chiptrack</li>
    <li>mullvad</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 3 packages built:</summary>
  <ul>
    <li>alfis-nogui</li>
    <li>conduwuit</li>
    <li>workstyle</li>
  </ul>
</details>

**Comment by niklaskorz** [comment]
Not sure if you intend to add more packages and which, but for e.g. zed-editor I migrated to the new fetcher in #356757 and it works fine.

**Comment by TomaSajt** [comment]
There are still some kinks to iron out, but the submodule stuff should have been the last change that could actually change the hash.

Just opened https://github.com/NixOS/nixpkgs/pull/356821 which decreases the concurrency to some more sane values.

**Comment by JohnRTitor** [comment]
@ofborg build alfis workstyle mullvad chiptrack cloud-hypervisor conduwuit xdg-desktop-portal-cosmic

**Comment by JohnRTitor** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 356385`

---
### `x86_64-linux`
<details>
  <summary>:x: 4 packages failed to build:</summary>
  <ul>
    <li>cloud-hypervisor</li>
    <li>cloud-hypervisor.debug</li>
    <li>conduwuit</li>
    <li>mullvad</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 7 packages built:</summary>
  <ul>
    <li>alfis</li>
    <li>alfis-nogui</li>
    <li>chiptrack</li>
    <li>cosmic-session</li>
    <li>workstyle</li>
    <li>xdg-desktop-portal-cosmic</li>
    <li>xdg-desktop-portal-cosmic.debug</li>
  </ul>
</details>

**Comment by JohnRTitor** [comment]
> Not sure if you intend to add more packages and which

I do not intend to migrate all packages with a Cargo.lock in nixpkgs in this PR, in fact I think it's just more rebuild work, and the packages should be either mass migrated, or migrated when updating.

The packages listed here were decided randomly, and my goal was to find and iron out the bugs before the mass migration and adoption by other packages.

**Comment by JohnRTitor** [comment]
@TomaSajt when running nixpkgs-review.

```log
┃        > The above exception was the direct cause of the following exception:
┃        >
┃        > Traceback (most recent call last):
┃        >   File "/nix/store/rhm7zn3pxrm6805v5fkbvyrxa53yf8fb-fetch-cargo-vendor-util/bin/fetch-cargo-vendor-util", line 283, in <module>
┃        >     main()
┃        >   File "/nix/store/rhm7zn3pxrm6805v5fkbvyrxa53yf8fb-fetch-cargo-vendor-util/bin/fetch-cargo-vendor-util", line 279, in main
┃        >     subcommand_func()
┃        >   File "/nix/store/rhm7zn3pxrm6805v5fkbvyrxa53yf8fb-fetch-cargo-vendor-util/bin/fetch-cargo-vendor-util", line 270, in <lambda>
┃        >     "create-vendor-staging": lambda: create_vendor_staging(lockfile_path=Path(sys.argv[2]), out_dir=Path(sys.argv[3])),
┃        >                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
┃        >   File "/nix/store/rhm7zn3pxrm6805v5fkbvyrxa53yf8fb-fetch-cargo-vendor-util/bin/fetch-cargo-vendor-util", line 133, in create_vendor_staging
┃        >     pool.starmap(download_tarball, tarball_args_gen)
┃        >   File "/nix/store/px2nj16i5gc3d4mnw5l1nclfdxhry61p-python3-3.12.7/lib/python3.12/multiprocessing/pool.py", line 375, in starmap
┃        >     return self._map_async(func, iterable, starmapstar, chunksize).get()
┃        >            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
┃        >   File "/nix/store/px2nj16i5gc3d4mnw5l1nclfdxhry61p-python3-3.12.7/lib/python3.12/multiprocessing/pool.py", line 774, in get
┃        >     raise self._value
┃        > requests.exceptions.ConnectionError: None: Max retries exceeded with url: /crates/rand_chacha/rand_chacha-0.3.1.crate (Caused by None)
┃        For full logs, run 'nix log /nix/store/vxnqpkg5h0x1sq63lp0x4iy0m5gn8pkd-cloud-hypervisor-42.0-vendor-staging.drv'.
```

**Comment by JohnRTitor** [comment]
There was a hash mismatch with cloud-hypervisor. Fixed.
@ofborg build alfis workstyle mullvad chiptrack cloud-hypervisor conduwuit xdg-desktop-portal-cosmic

**Comment by TomaSajt** [comment]
> There was a hash mismatch with cloud-hypervisor. Fixed. @ofborg build alfis workstyle mullvad chiptrack cloud-hypervisor conduwuit xdg-desktop-portal-cosmic

I guess you forgot to reset your branch to the upstream one, so you accidentally overrode my fixed hash with the outdated one when rebasing.
But it should be okay now.

**Comment by JohnRTitor** [comment]
@TomaSajt hypervisor and mullvad is failing to be built in nixpkgs-review (ran after garbage collecting) due to maximum no of retries hit (not because of a hash error)

https://termbin.com/3mz8

**Comment by TomaSajt** [comment]
You could add a temporary commit on top of the PR that increases the retry count, like shown before.
And then use nixpkgs-review rev HEAD

**Comment by JohnRTitor** [comment]
Gonna merge this, since eval passed along with a nixpkgs-review

**Pull request merged**